python/ray/dag/dag_node.py

```diff
@@ -58,12 +58,30 @@ def __init__(
         self._bound_other_args_to_resolve: Optional[Dict[str, Any]] = (
             other_args_to_resolve or {}
         )
+
+        # The list of nodes that use this DAG node as input.
+        self._downstream_nodes: List[DAGNode] = []
+        # The list of nodes that this DAG node uses as input.
+        scanner = _PyObjScanner()
+        self._upstream_nodes: List[DAGNode] = scanner.find_nodes(
+            [
+                self._bound_args,
+                self._bound_kwargs,
+                self._bound_other_args_to_resolve,
+            ]
+        )
+        scanner.clear()
+        # Update upstream dependencies.
+        for upstream_node in self._upstream_nodes:
+            upstream_node._downstream_nodes.append(self)
```

Comment:
```suggestion
            upstream_node._downstream_nodes.append(self)
```
consider introducing an API like update_downstream_node!

Reply:
Updated: https://github.com/ray-project/ray/pull/47261/commits/54df435b62b4510d731bd9df4b766f694e54d5ea